### ISSUE TITLE
Related to request Timeout

### ISSUE THREAD

**Issue by rohit-raj** [firebase/flashlight#69]
Though it is good that we keep requestTimeout while we create an elastic client instance. When I do not keep this parameter, then my indexer stops and throws :
**failed to index indexname/resource/resourcename: Error: Request Timeout after 30000ms**
by which my most of the firebase data are not indexed.

Ideally, the requestTimeout is a parameter for when the request is taking too long then timeout that request, but in my case the request is always fulfilled though the indexer timeouts after 30 seconds.

my indexed data flows like
**indexed indexname/resource/0000df7b-b9ea-4496-b2e6-bb9c86aa8b5c
indexed indexname/resource/00008527-5f7f-43c2-97c8-bd71c6d409ca
indexed indexname/resource/000cd4b8-5dba-4614-b699-dfb592490ca4
indexed indexname/resource/00014bd1-88dc-4ff4-8248-ef4d071ba28d
indexed indexname/resource/0016f0a2-edda-4d2a-9ea6-d8c6e9575b4e
indexed indexname/resource/0018fe53-1e45-4925-9580-e3851ee9728e
indexed indexname/resource/0012b89d-10a5-4d54-9134-453e87fdc67c
indexed indexname/resource/00059da0-5174-4c48-ab3c-fd342c4edbf4**

after 30 seconds:
**failed to index indexname/resource/ffd5517e-cab0-4bec-8989-4f6d760f18eb: Error: Request Timeout after 30000ms
failed to index indexname/resource/ffdb0ad0-c9bb-4ebd-9f30-9aecf4ffedd3: Error: Request Timeout after 30000ms
failed to index indexname/resource/ffdfce2e-e11e-415c-b9f6-8e1f6e6777d7: Error: Request Timeout after 30000ms
failed to index indexname/resource/ffe39ae2-ad3e-4944-a65c-744490407247: Error: Request Timeout after 30000ms
failed to index indexname/resource/fff3c43f-6082-4cad-97ab-bfe04c61039a: Error: Request Timeout after 30000ms**

in config js i have:
**exports.paths = [
   {
      path:  "resources",
      index: "indexname",
      type:  "resource"
   }
];**

my firebase data looks like:
![firebase](https://cloud.githubusercontent.com/assets/11807587/17401141/93cece96-5a69-11e6-8a9d-0f39919d5d1a.png)

is there any problem with my data or the structure of the firebase data storage, or is there any other thing that I need to take care while indexing ?

Since the current data that I am indexing is in dev mode, then its not the issue for me keeping the requestTimeout to any number that is sufficient to index all the data, but my production database is quiet huge I cannot decide on what my requestTimeout should be inorder to index all the data.

Please let me know any leads on the above issue.

Secondly, is this only designed to run as a service, or can we alter it to index the data then exit from the program and then run the update or delete part to just index only the new changes in firebase (maybe new record created, old record updated, or deleted)? but not touching all the already indexed data.

**Comment by katowulf** [comment]
Please provide a minimal repro with sanitized sample data that recreates the conditions you're describing. I'll be happy to take a look. Alternately, if you already know where the problem is, feel free to submit a PR and a test unit to address this.

**Comment by rohit-raj** [comment]
For reproducing this error, please follow the below steps:
1> Extract and import the attached file to your firebase database.
2> Get the flashlight code in your system.
3> Update the config with the dbUrl(where you imported the step 1 data) and serviceAccount.
2> Run app.js to index the imported data(from step 1) of your firebase db.
3> The error will be reproduced.

config.js:
```
exports.paths = [
   {
        path: "resources",
        index: "indexname",
        type: "resource"
   }
];
```
Do let me know your doubts if any.

Test database json:
[resources.tar.gz](https://github.com/firebase/flashlight/files/403253/resources.tar.gz)

**Comment by rohit-raj** [comment]
Hi,
Were you able to reproduce the error ?
Please let me know if you need any other info.

**Comment by Jason-CloudTP** [comment]
I have this same issue with a large parts inventory.  Is there resolution yet?

**Comment by rohit-raj** [comment]
the solution is not available till date, I am trying to put the requestTimeout to a large value to avoid connection timeout..
I am in middle of some other development, hence I am not able to give time to this. 
There are two different things to implement:
1> The timeout for large chunk of data
2> resuming the indexer from a point at which it was previously stopped.

**Comment by katowulf** [comment]
I'm going to close this as resolved by commit dc5e3b4b1473b8c491daf922367e79c281486abd. I think that using the pathBuilder() and timestamps to limit initial load volume will probably fix errors like these (related to big data sets that can't be indexed easily) and be a decent workaround for now. Additional work from #74 will eventually improve this even more.
